fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/DimensionIdsFixer.java

```diff
@@ -0,0 +1,119 @@
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraft.world.level.LevelProperties;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Consumer;
+
+/**
+ * Handles fixing raw dimension ids between saves and servers
+ */
+public class DimensionIdsFixer {
+    static final Identifier ID = new Identifier("fabric", "dimension/sync");
+
+    /**
+     * Assigns a unique id to every registered {@link FabricDimensionType}, keeping the known ids
+     * from {@code savedIds}.
+     * @param savedIds a compound tag mapping dimension ids to raw ids
+     * @return id to raw id mappings of the current instance
+     * @throws RemapException if dimensions IDs conflict irredeemably
+     */
+    public static CompoundTag apply(CompoundTag savedIds) throws RemapException {
+        /*
+         * We want to give to each fabric dimension a unique ID. We also want to give back previously assigned ids.
+         * And we have to take into account non-fabric dimensions, which raw IDs cannot change.
+         * So we iterate over every dimension, note the ones which id cannot change, then update the free ones.
+         */
+        Int2ObjectMap<Identifier> fixedIds = new Int2ObjectOpenHashMap<>();
+        List<FabricDimensionType> fabricDimensions = new ArrayList<>();
+        CompoundTag fabricDimensionIds = new CompoundTag();
+        // step 1: detect all fabric and non-fabric dimensions
+        for (Identifier id : Registry.DIMENSION.getIds()) {
+            DimensionType dimensionType = Objects.requireNonNull(DimensionType.byId(id));
+            if (dimensionType instanceof FabricDimensionType) {
+                fabricDimensions.add((FabricDimensionType) dimensionType);
+            } else {
+                Identifier existing = fixedIds.put(dimensionType.getRawId(), id);
+                if (existing != null) {
+                    throw new RemapException("Two non-fabric dimensions have the same raw dim id (" + dimensionType.getRawId() + ") : " + existing + " and " + id);
+                }
+            }
+        }
+        // step 2: read saved ids
+        for (String key : savedIds.getKeys()) {
+            int savedRawId = savedIds.getInt(key);
+            Identifier dimId = new Identifier(key);
+            Identifier existing = fixedIds.putIfAbsent(savedRawId, dimId);
+            if (existing != null && !existing.equals(dimId)) {
+                throw new RemapException("Saved fabric dimension got replaced with a non-fabric one! " + dimId + " replaced with " + existing + " (raw id: " + savedRawId + ")");
+            }
+            DimensionType dim = DimensionType.byId(dimId);
+            if (dim instanceof FabricDimensionType) {
+                ((FabricDimensionType) dim).setFixedId(savedRawId);
```

Comment:
This can make it not-so-fixed if set by a mod previously, I think it should either be made more obvious that it's a "desired id" or abort on mismatch. This internal "id setting" (probably) shouldn't share the same method+field as the mod-side id setup.

Reply:
Right, so there should be a "desired raw id" that can be set by a mod in `FabricDimensionType`, and a separate internal raw id setting that is set by this method. How should the setter for such an internal id be handled ? Should it be invisible to API consumers, with a lambda access being passed to the internal classes, or should it be a regular public method with big warnings in the javadoc ?

Reply:
I think the best choice for now is to set the internal id from impl via reflection, it's simple and doesn't distort the api class like other approaches.